linux/devdoc/completion_port_linux_requirements.md

```diff
@@ -0,0 +1,187 @@
+# completion_port_linux requirements
+
+## Overview
+
+The completion_port_linux module handles the threading and control of the epoll socket system
+
+## Exposed API
+
+```C
+typedef struct COMPLETION_PORT_TAG* COMPLETION_PORT_HANDLE;
+
+#define COMPLETION_PORT_EPOLL_ACTION_VALUES \
+    COMPLETION_PORT_EPOLL_EPOLLRDHUP, \
+    COMPLETION_PORT_EPOLL_EPOLLIN, \
+    COMPLETION_PORT_EPOLL_EPOLLOUT, \
+    COMPLETION_PORT_CLOSING
+
+MU_DEFINE_ENUM(COMPLETION_PORT_EPOLL_ACTION, COMPLETION_PORT_EPOLL_ACTION_VALUES)
+
+#define COMPLETION_PORT_EPOLL_RESULT_VALUES \
+    COMPLETION_PORT_RESULT_OK, \
+    COMPLETION_PORT_RESULT_ERROR, \
+    COMPLETION_PORT_RESULT_ABANDONED
+
+MU_DEFINE_ENUM(COMPLETION_PORT_RESULT, COMPLETION_PORT_RESULT_VALUES)
+
+typedef void (*ON_COMPLETION_PORT_EVENT_COMPLETE)(void* context, COMPLETION_PORT_EPOLL_ACTION epoll_action, COMPLETION_PORT_RESULT result, int32_t amount_transfered);
+
+typedef struct EPOLL_BUFFER_DATA_TAG
+{
+    void* buffer;
+    uint32_t length;
+} EPOLL_BUFFER_DATA;
+
+typedef struct COMPLETION_PORT_INGRESS_DATA_TAG
+{
+    int epoll_op;
+    SOCKET_HANDLE socket;
+    ON_COMPLETION_PORT_EVENT_COMPLETE event_callback;
+    void* event_callback_ctx;
+    EPOLL_BUFFER_DATA buffer_data;
+} COMPLETION_PORT_INGRESS_DATA;
+
+MOCKABLE_FUNCTION(, COMPLETION_PORT_HANDLE, completion_port_create);
+MOCKABLE_FUNCTION(, void, completion_port_inc_ref, COMPLETION_PORT_HANDLE, completion_port);
+MOCKABLE_FUNCTION(, void, completion_port_dec_ref, COMPLETION_PORT_HANDLE, completion_port);
+MOCKABLE_FUNCTION(, int, completion_port_add, COMPLETION_PORT_HANDLE, completion_port, const COMPLETION_PORT_INGRESS_DATA*, ingress_data);
```

Comment:
>const COMPLETION_PORT_INGRESS_DATA*, ingress_data [](http://example.com/codeflow?start=87&length=49)

Can we have the parameters unfolded in the API (it makes it much harder for testing to have it as a struct.